nifi-nar-bundles/nifi-questdb-bundle/nifi-questdb/src/test/java/org/apache/nifi/questdb/embedded/EmbeddedDatabaseManagerTest.java

```diff
@@ -0,0 +1,397 @@
+import static org.apache.nifi.questdb.util.QuestDbTestUtil.SELECT_QUERY;
+import static org.apache.nifi.questdb.util.QuestDbTestUtil.getRandomTestData;
+import static org.apache.nifi.questdb.util.QuestDbTestUtil.getTestData;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertIterableEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class EmbeddedDatabaseManagerTest extends EmbeddedQuestDbTest {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EmbeddedDatabaseManagerTest.class);
+    private static final int DAYS_TO_KEEP_EVENT = 1;
+    private static final String NON_EXISTING_PLACE = SystemUtils.IS_OS_WINDOWS ? "T:/nonExistingPlace" : "/nonExistingPlace";
+
+    @Test
+    public void testAcquiringWithoutInitialization() {
+        final EmbeddedDatabaseManager testSubject = new EmbeddedDatabaseManager(new SimpleEmbeddedDatabaseManagerContext());
+        assertThrows(IllegalStateException.class, () -> testSubject.acquireClient());
+    }
+
+    @Test
+    public void testHappyPath() throws DatabaseException {
+        final List<Event> testData = getTestData();
+        final DatabaseManager testSubject = getTestSubject();
+        assertDatabaseFolderIsNotEmpty();
+
+        final Client client = testSubject.acquireClient();
+
+        client.insert(EVENT_TABLE_NAME, QuestDbTestUtil.getEventTableDataSource(testData));
+        final List<Event> result = client.query(SELECT_QUERY, RequestMapping.getResultProcessor(QuestDbTestUtil.EVENT_TABLE_REQUEST_MAPPING));
+
+        assertIterableEquals(testData, result);
+
+        testSubject.close();
+
+        // Even if the client itself is not connected, manager prevents client to reach database after closing
+        assertFalse(client.query(SELECT_QUERY, RequestMapping.getResultProcessor(QuestDbTestUtil.EVENT_TABLE_REQUEST_MAPPING)).iterator().hasNext());
+    }
+
+    @Test
+    public void testRollover() throws DatabaseException, InterruptedException {
+        final List<Event> testData = new ArrayList<>();
+        testData.add(new Event(Instant.now().minus((DAYS_TO_KEEP_EVENT + 1), ChronoUnit.DAYS), "A", 1));
+        testData.add(new Event(Instant.now(), "B", 2));
+        final DatabaseManager testSubject = getTestSubject();
+
+        final Client client = testSubject.acquireClient();
+        client.insert(EVENT_TABLE_NAME, QuestDbTestUtil.getEventTableDataSource(testData));
+
+        // The rollover runs in every 5 seconds
+        Thread.sleep(TimeUnit.SECONDS.toMillis(6));
+
+        final List<Event> result = client.query(SELECT_QUERY, RequestMapping.getResultProcessor(QuestDbTestUtil.EVENT_TABLE_REQUEST_MAPPING));
+        testSubject.close();
+
+        assertEquals(1, result.size());
```

Comment:
When building, the module, this test method fails intermittently. It seems like the `Thread.sleep()` approach is not a strong enough guarantee of rollover completion. It may be necessary to do some additional checking on the directory content itself to determine if a rollover has occurred. If all else fails, it would be better to remove the test method, as opposed to having a test that fails intermittently.

```
[ERROR] Tests run: 13, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 18.24 s <<< FAILURE! -- in org.apache.nifi.questdb.embedded.EmbeddedDatabaseManagerTest
[ERROR] org.apache.nifi.questdb.embedded.EmbeddedDatabaseManagerTest.testRollover -- Time elapsed: 7.888 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <1> but was: <2>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:145)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:527)
	at org.apache.nifi.questdb.embedded.EmbeddedDatabaseManagerTest.testRollover(EmbeddedDatabaseManagerTest.java:109)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

Reply:
This is an answer for both this and the previous one. 

In general I would lean toward to use something like a latch, which guarantees the end of the background task. Due to the nature of the test case I could not use a spy for this, and I rely on waiting. As I left some gap between the roundtrip and wait time, I did not have issue with it so far, but based on your concerns I proposed a more refined way to do this. The fundamentals are the same: the test thread waits, but it polls multiple times to shorten the usual test duration and now it has 2 sec of gaps between the scheduled round trips and the waiting times. In case 3 seconds is not enough for the positive result I assume we have an issue indeed. Either with the test or the environment

Reply:
Thanks for adjusting the approach, the polling strategy looks good, and avoids the timing issues.